File: samples/dnn/granite_docling_inference.cpp
// This file is part of OpenCV project.
// It is subject to the license terms in the LICENSE file found in the top-level directory
// of this distribution and at http://opencv.org/license.html.

/*
 * This is a sample program to run Granite-Docling-258M vision-language inference in OpenCV
 * using ONNX models and OpenCV's new DNN engine (ENGINE_NEW). Given a page image and a
 * text prompt, it generates a "doctag" style text response describing the page (OCR text,
 * tables, form fields, and layout/section structure).
 *
 * The model is split into three ONNX files:
 *     - Vision encoder : image tiles -> image features
 *     - Embed tokens    : prompt token ids -> text embeddings
 *     - Decoder         : [image features | text embeddings] -> logits (with KV-cache)
 *
 * Model: https://huggingface.co/ibm-granite/granite-docling-258M
 * ONNX:  https://huggingface.co/onnx-community/granite-docling-258M-ONNX
 *
 * Run the sample:
 * 1. Download the plain (non-quantized) fp32 ONNX export into <model_dir>, keeping the
 *    upstream layout (config.json, preprocessor_config.json, processor_config.json,
 *    tokenizer.json at the root; onnx/vision_encoder.onnx, onnx/embed_tokens.onnx,
 *    onnx/decoder_model_merged.onnx under onnx/).
 *
 * 2. Run:
 *
 *      ./granite_docling_inference --model_dir=<model_dir> \
 *                                  --input=<path-to-page-image>
 *
 */

#include <algorithm>
#include <cmath>
#include <cstring>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

#include <opencv2/core.hpp>
#include <opencv2/dnn.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

using namespace cv;
using namespace cv::dnn;
using namespace std;

static const string DEFAULT_PROMPT =
    "Convert this page to docling. Preserve OCR text, table structure, "
    "form fields, and layout/section structure.";

static void resizeAA(const Mat& src, Mat& dst, Size size)
{
    bool shrinking = (int64_t)size.width * size.height < (int64_t)src.cols * src.rows;
    resize(src, dst, size, 0, 0, shrinking ? INTER_AREA : INTER_LANCZOS4);
}

static Mat tileImage(const Mat& imageBgr, int longestEdge, int tileSize,
                     const Vec3f& mean, const Vec3f& std_, int& rowsOut, int& colsOut)
{
    int h0 = imageBgr.rows, w0 = imageBgr.cols;
    int newW, newH;
    if (w0 >= h0)
    {
        newW = longestEdge;
        newH = max(1, (int)round((double)longestEdge * h0 / w0));
    }
    else
    {
        newH = longestEdge;
        newW = max(1, (int)round((double)longestEdge * w0 / h0));
    }

    Mat resized;
    resizeAA(imageBgr, resized, Size(newW, newH));

    int rows = (newH + tileSize - 1) / tileSize;
    int cols = (newW + tileSize - 1) / tileSize;
    rowsOut = rows;
    colsOut = cols;

    Mat grid;
    resizeAA(resized, grid, Size(cols * tileSize, rows * tileSize));

    int numTiles = rows * cols + 1;
    int sizes[] = {1, numTiles, 3, tileSize, tileSize};
    Mat pixelValues(5, sizes, CV_32F);

    auto normalizeTile = [&](const Mat& tileBgr, int tileIdx)
    {
        Mat tileRgb;
        cvtColor(tileBgr, tileRgb, COLOR_BGR2RGB);
        tileRgb.convertTo(tileRgb, CV_32F, 1.0 / 255.0);
        vector<Mat> channels(3);
        split(tileRgb, channels);
        for (int c = 0; c < 3; c++)
        {
            Mat dst(tileSize, tileSize, CV_32F, pixelValues.ptr<float>(0, tileIdx, c));
            channels[c].convertTo(dst, -1, 1.0 / std_[c], -mean[c] / std_[c]);
        }
    };

    int idx = 0;
    for (int r = 0; r < rows; r++)
        for (int c = 0; c < cols; c++)
        {
            Rect roi(c * tileSize, r * tileSize, tileSize, tileSize);
            normalizeTile(grid(roi), idx++);
        }
    Mat thumbnail;
    resizeAA(resized, thumbnail, Size(tileSize, tileSize));
    normalizeTile(thumbnail, idx);

    return pixelValues;
}

static string buildPrompt(int rows, int cols, int imageSeqLen, const string& userText)
{

    ostringstream imagePart;
    for (int h = 0; h < rows; h++)
    {
        for (int w = 0; w < cols; w++)
        {
            imagePart << "<fake_token_around_image><row_" << (h + 1) << "_col_" << (w + 1) << ">";
            for (int i = 0; i < imageSeqLen; i++)
                imagePart << "<image>";
        }
        imagePart << "\n";
    }
    imagePart << "\n<fake_token_around_image><global-img>";
    for (int i = 0; i < imageSeqLen; i++)
        imagePart << "<image>";
    imagePart << "<fake_token_around_image>";

    ostringstream full;
    full << "<|start_of_role|>user<|end_of_role|>" << imagePart.str() << userText
         << "<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>";
    return full.str();
}

static int argmaxLastToken(const Mat& logits)
{
    int seqLen = logits.size[1];
    int vocabSize = logits.size[2];
    const float* row = logits.ptr<float>(0, seqLen - 1);
    return (int)(max_element(row, row + vocabSize) - row);
}

static vector<int> graniteDoclingInference(Net& visionNet, Net& embedNet, Net& decoderNet,
                                           const Mat& pixelValues, const string& prompt,
                                           int maxNewTokens, Tokenizer& tokenizer,
                                           int imageTokenId, int eosTokenId)
{
    cout << "Inferencing Granite-Docling-258M model..." << endl;

    vector<int> tokens = tokenizer.encode(prompt);
    int promptLen = (int)tokens.size();
    vector<int64_t> inputIdsData(tokens.begin(), tokens.end());
    int idsShape[] = {1, promptLen};
    Mat inputIds(2, idsShape, CV_64S, inputIdsData.data());

    int maskShape[] = {1, pixelValues.size[1], pixelValues.size[3], pixelValues.size[4]};
    Mat pixelAttentionMask(4, maskShape, CV_Bool, Scalar(1));

    visionNet.setInput(pixelValues, "pixel_values");
    visionNet.setInput(pixelAttentionMask, "pixel_attention_mask");
    Mat imageFeatures = visionNet.forward();

    embedNet.setInput(inputIds, "input_ids");
    Mat inputsEmbeds = embedNet.forward();

    int hiddenDim = inputsEmbeds.size[2];
    float* embedsData = inputsEmbeds.ptr<float>();
    const float* featData = imageFeatures.ptr<float>();
    int featIdx = 0;
    for (int i = 0; i < promptLen; i++)
        if (tokens[i] == imageTokenId)
            memcpy(embedsData + (size_t)i * hiddenDim, featData + (size_t)(featIdx++) * hiddenDim, hiddenDim * sizeof(float));

    decoderNet.enableKVCache();
    vector<int64_t> maskData(promptLen, 1);
    int attnShape[] = {1, promptLen};
    Mat attentionMask(2, attnShape, CV_64S, maskData.data());

    decoderNet.setInput(inputsEmbeds, "inputs_embeds");
    decoderNet.setInput(attentionMask, "attention_mask");
    Mat logits = decoderNet.forward();
    int newId = argmaxLastToken(logits);
    vector<int> generated = {newId};

    for (int step = 0; step < maxNewTokens - 1; step++)
    {
        if (newId == eosTokenId)
            break;

        int64_t idData[1] = {newId};
        int idShape[] = {1, 1};
        Mat newIdMat(2, idShape, CV_64S, idData);
        embedNet.setInput(newIdMat, "input_ids");
        Mat newEmbed = embedNet.forward();

        maskData.push_back(1);
        int newAttnShape[] = {1, (int)maskData.size()};
        Mat newAttentionMask(2, newAttnShape, CV_64S, maskData.data());

        decoderNet.setInput(newEmbed, "inputs_embeds");
        decoderNet.setInput(newAttentionMask, "attention_mask");
        logits = decoderNet.forward();
        newId = argmaxLastToken(logits);
        generated.push_back(newId);
    }

    if (!generated.empty() && generated.back() == eosTokenId)
        generated.pop_back();

    return generated;
}

int main(int argc, char** argv)
{
    const string keys =
        "{ help h         |      | Print help message }"
        "{ model_dir      |      | Path to the local onnx-community/granite-docling-258M-ONNX export "
        "(config.json, preprocessor_config.json, processor_config.json, "
        "onnx/vision_encoder.onnx, onnx/embed_tokens.onnx, onnx/decoder_model_merged.onnx) }"
        "{ input i        |      | Path to the input page image }"
        "{ prompt         |      | Task prompt (defaults to the built-in docling conversion prompt) }"
        "{ max_new_tokens | 512  | Maximum number of new tokens to generate }"
        "{ seed           | 0    | Random seed }";

    CommandLineParser parser(argc, argv, keys);
    parser.about("Use this sample to run Granite-Docling-258M vision-language inference in OpenCV");
    if (parser.has("help") || !parser.has("model_dir") || !parser.has("input"))
    {
        parser.printMessage();
        return 0;
    }

    string modelDir = parser.get<String>("model_dir");
    string inputPath = parser.get<String>("input");
    string prompt = parser.get<String>("prompt");
    if (prompt.empty())
        prompt = DEFAULT_PROMPT;
    int maxNewTokens = parser.get<int>("max_new_tokens");
    setRNGSeed(parser.get<int>("seed"));

    cout << "Preparing Granite-Docling-258M model..." << endl;
    Tokenizer tokenizer = Tokenizer::loadVLM(modelDir + "/", "granite-docling");

    FileStorage config(modelDir + "/config.json", FileStorage::READ | FileStorage::FORMAT_JSON);
    if (!config.isOpened())
    {
        cerr << "Could not open " << modelDir + "/config.json" << endl;
        return 1;
    }
    FileStorage preprocessor(modelDir + "/preprocessor_config.json", FileStorage::READ | FileStorage::FORMAT_JSON);
    if (!preprocessor.isOpened())
    {
        cerr << "Could not open " << modelDir + "/preprocessor_config.json" << endl;
        return 1;
    }
    FileStorage processor(modelDir + "/processor_config.json", FileStorage::READ | FileStorage::FORMAT_JSON);
    if (!processor.isOpened())
    {
        cerr << "Could not open " << modelDir + "/processor_config.json" << endl;
        return 1;
    }

    int imageTokenId = (int)config["image_token_id"];
    FileNode textConfig = config["text_config"];
    int eosTokenId = (!textConfig.empty() && !textConfig["eos_token_id"].empty())
                        ? (int)textConfig["eos_token_id"]
                        : (int)config["eos_token_id"];
    int imageSeqLen = (int)processor["image_seq_len"];
    int longestEdge = (int)preprocessor["size"]["longest_edge"];
    int maxTileEdge = (int)preprocessor["max_image_size"]["longest_edge"];
    vector<float> meanVec, stdVec;
    preprocessor["image_mean"] >> meanVec;
    preprocessor["image_std"] >> stdVec;
    Vec3f mean(meanVec[0], meanVec[1], meanVec[2]);
    Vec3f stdDev(stdVec[0], stdVec[1], stdVec[2]);

    Net visionNet  = readNetFromONNX(modelDir + "/onnx/vision_encoder.onnx", ENGINE_NEW);
    Net embedNet   = readNetFromONNX(modelDir + "/onnx/embed_tokens.onnx", ENGINE_NEW);
    Net decoderNet = readNetFromONNX(modelDir + "/onnx/decoder_model_merged.onnx", ENGINE_NEW);

    Mat image = imread(inputPath);
    if (image.empty())
    {
        cerr << "Could not read image: " << inputPath << endl;
        return 1;
    }

    int rows, cols;
    Mat pixelValues = tileImage(image, longestEdge, maxTileEdge, mean, stdDev, rows, cols);
    string fullPrompt = buildPrompt(rows, cols, imageSeqLen, prompt);
    cout << "Prompt:\n" << prompt << endl;

    vector<int> generated = graniteDoclingInference(visionNet, embedNet, decoderNet, pixelValues, fullPrompt,
                                                     maxNewTokens, tokenizer, imageTokenId, eosTokenId);
    string response = tokenizer.decode(generated);
    cout << "Response:\n" << response << endl;
    return 0;
}

File: samples/dnn/granite_docling_inference.py
# This file is part of OpenCV project.
# It is subject to the license terms in the LICENSE file found in the top-level directory
# of this distribution and at http://opencv.org/license.html.
# Copyright (C) 2026, BigVision LLC, all rights reserved.
# Third party copyrights are property of their respective owners.

'''
This is a sample script to run Granite-Docling-258M vision-language inference in OpenCV
using ONNX models and OpenCV's new DNN engine (ENGINE_NEW). Given a page image and a
text prompt, it generates a "doctag" style text response describing the page (OCR text,
tables, form fields, and layout/section structure).

The model is split into three ONNX files:
    - Vision encoder : image tiles -> image features
    - Embed tokens    : prompt token ids -> text embeddings
    - Decoder         : [image features | text embeddings] -> logits (with KV-cache)

Model: https://huggingface.co/ibm-granite/granite-docling-258M
ONNX:  https://huggingface.co/onnx-community/granite-docling-258M-ONNX

Run the script:
1. Download the plain (non-quantized) fp32 ONNX export into <model_dir>, keeping the
   upstream layout (config.json, preprocessor_config.json, processor_config.json,
   tokenizer.json at the root; onnx/vision_encoder.onnx, onnx/embed_tokens.onnx,
   onnx/decoder_model_merged.onnx under onnx/):

2. Run the script:

    python granite_docling_inference.py --model_dir=<model_dir> \
                                        --input=<path-to-page-image>

'''

import math
import json
import argparse
import numpy as np
import cv2 as cv

DEFAULT_PROMPT = ('Convert this page to docling. Preserve OCR text, table structure, '
                   'form fields, and layout/section structure.')

def parse_args():
    parser = argparse.ArgumentParser(description='Use this script to run Granite-Docling-258M vision-language inference in OpenCV',
                                    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
    parser.add_argument('--model_dir', type=str, required=True,
                        help='Path to the local onnx-community/granite-docling-258M-ONNX export '
                             '(config.json, preprocessor_config.json, processor_config.json, '
                             'onnx/{vision_encoder,embed_tokens,decoder_model_merged}.onnx).')
    parser.add_argument('--input', '-i', type=str, required=True, help='Path to the input page image.')
    parser.add_argument('--prompt', type=str, default=DEFAULT_PROMPT, help='Task prompt.')
    parser.add_argument('--max_new_tokens', type=int, default=512, help='Maximum number of new tokens to generate.')
    parser.add_argument('--seed', type=int, default=0, help='Random seed.')
    return parser.parse_args()

def load_json(path):
    with open(path, 'r', encoding='utf-8') as f:
        return json.load(f)

def resize(image, size):
    dst_w, dst_h = size
    src_h, src_w = image.shape[:2]
    shrinking = dst_w * dst_h < src_w * src_h
    return cv.resize(image, size, interpolation=cv.INTER_AREA if shrinking else cv.INTER_LANCZOS4)

def tile_image(image_bgr, longest_edge, tile_size, mean, std):
    h0, w0 = image_bgr.shape[:2]
    if w0 >= h0:
        new_w, new_h = longest_edge, max(1, round(longest_edge * h0 / w0))
    else:
        new_w, new_h = max(1, round(longest_edge * w0 / h0)), longest_edge
    resized = resize(image_bgr, (new_w, new_h))

    rows, cols = math.ceil(new_h / tile_size), math.ceil(new_w / tile_size)
    mean_arr = np.asarray(mean, dtype=np.float32)
    std_arr = np.asarray(std, dtype=np.float32)

    def normalize(tile_bgr):
        tile_rgb = cv.cvtColor(tile_bgr, cv.COLOR_BGR2RGB).astype(np.float32) / 255.0
        return ((tile_rgb - mean_arr) / std_arr).transpose(2, 0, 1)  # HWC -> CHW

    grid = resize(resized, (cols * tile_size, rows * tile_size))
    pixel_values = np.zeros((rows * cols + 1, 3, tile_size, tile_size), dtype=np.float32)
    idx = 0
    for r in range(rows):
        for c in range(cols):
            y0, x0 = r * tile_size, c * tile_size
            pixel_values[idx] = normalize(grid[y0:y0 + tile_size, x0:x0 + tile_size])
            idx += 1
    thumbnail = resize(resized, (tile_size, tile_size))
    pixel_values[idx] = normalize(thumbnail)
    return pixel_values[np.newaxis], rows, cols

def build_prompt(rows, cols, image_seq_len, user_text):
    image_part = ''
    for h in range(rows):
        for w in range(cols):
            image_part += f'<fake_token_around_image><row_{h + 1}_col_{w + 1}>' + '<image>' * image_seq_len
        image_part += '\n'
    image_part += '\n<fake_token_around_image><global-img>' + '<image>' * image_seq_len + '<fake_token_around_image>'
    return f'<|start_of_role|>user<|end_of_role|>{image_part}{user_text}<|end_of_text|>\n<|start_of_role|>assistant<|end_of_role|>'

def granite_docling_inference(vision_net, embed_net, decoder_net, pixel_values, prompt, max_new_tokens, tokenizer, image_token_id, eos_token_id):

    print('Inferencing Granite-Docling-258M model...')

    input_ids = np.array([tokenizer.encode(prompt)], dtype=np.int64)
    prompt_len = input_ids.shape[1]

    vision_net.setInput(pixel_values, 'pixel_values')
    vision_net.setInput(np.ones(pixel_values.shape[:2] + pixel_values.shape[3:], dtype=bool), 'pixel_attention_mask')
    image_features = vision_net.forward()

    embed_net.setInput(input_ids, 'input_ids')
    inputs_embeds = embed_net.forward()
    inputs_embeds[input_ids == image_token_id] = image_features.reshape(-1, image_features.shape[-1])

    decoder_net.enableKVCache()
    attention_mask = np.ones((1, prompt_len), dtype=np.int64)

    decoder_net.setInput(inputs_embeds, 'inputs_embeds')
    decoder_net.setInput(attention_mask, 'attention_mask')
    logits = decoder_net.forward()
    new_id = int(np.argmax(logits[:, -1, :].reshape(-1)))
    generated = [new_id]

    for _ in range(max_new_tokens - 1):
        if new_id == eos_token_id:
            break
        embed_net.setInput(np.array([[new_id]], dtype=np.int64), 'input_ids')
        new_embed = embed_net.forward()
        attention_mask = np.concatenate([attention_mask, [[1]]], axis=-1)
        decoder_net.setInput(new_embed, 'inputs_embeds')
        decoder_net.setInput(attention_mask, 'attention_mask')
        logits = decoder_net.forward()
        new_id = int(np.argmax(logits[:, -1, :].reshape(-1)))
        generated.append(new_id)

    if generated and generated[-1] == eos_token_id:
        generated.pop()

    return generated

if __name__ == '__main__':

    args = parse_args()
    np.random.seed(args.seed)

    print('Preparing Granite-Docling-258M model...')
    tokenizer = cv.dnn.Tokenizer.loadVLM(args.model_dir + '/', 'granite-docling')

    config = load_json(args.model_dir + '/config.json')
    preprocessor = load_json(args.model_dir + '/preprocessor_config.json')
    image_seq_len = load_json(args.model_dir + '/processor_config.json')['image_seq_len']
    image_token_id = int(config['image_token_id'])
    eos_token_id = int(config.get('text_config', {}).get('eos_token_id', config.get('eos_token_id')))

    vision_net  = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/vision_encoder.onnx', cv.dnn.ENGINE_NEW)
    embed_net   = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/embed_tokens.onnx', cv.dnn.ENGINE_NEW)
    decoder_net = cv.dnn.readNetFromONNX(args.model_dir + '/onnx/decoder_model_merged.onnx', cv.dnn.ENGINE_NEW)

    image = cv.imread(args.input)
    if image is None:
        raise IOError('Could not read image: ' + args.input)
    pixel_values, rows, cols = tile_image(
        image,
        preprocessor['size']['longest_edge'],
        preprocessor['max_image_size']['longest_edge'],
        preprocessor['image_mean'],
        preprocessor['image_std'],
    )
    prompt = build_prompt(rows, cols, image_seq_len, args.prompt)
    print(f'Prompt:\n{args.prompt}')

    generated = granite_docling_inference(vision_net, embed_net, decoder_net, pixel_values, prompt,
                                          args.max_new_tokens, tokenizer, image_token_id, eos_token_id)
    response = tokenizer.decode(generated)
    print(f'Response:\n{response}')
